src/test/java/org/broadinstitute/hellbender/tools/FixCallSetSampleOrderingIntegrationTest.java

```diff
@@ -0,0 +1,116 @@
+                        .genotypes(new GenotypeBuilder(sampleName, alleles).attribute(SAMPLE_NAME_KEY, sampleName).make())
+                        .make();
+                writer.add(variant);
+                nameToFileMap.put(sampleName, out);
+            }
+        }
+
+        final File mapping = IOUtils.writeTempFile(
+                nameToFileMap.entrySet().stream()
+                        .map(pair -> pair.getKey() + "\t" + pair.getValue())
+                        .sorted(Comparator.reverseOrder())
+                        .collect(Collectors.joining("\n")), "mapping", ".sample_map");
+     //   mapping.deleteOnExit();
+        return mapping;
+    }
+
+    @Test(enabled = false)
+    public void generateVCFs() throws IOException {
+        final File file = writeManyVCFs(1000);
+        System.out.println(file);
+    }
+
+    @DataProvider
+    public Object[][] getBadlySortedFiles(){
+        return new Object[][]{
+                /**
+                 *   test files were created by generating mangled files using the {@link #writeManyVCFs(1000)
+                 *   these were then imported to GenomicsDB using GATK 4.beta.5-66-g9609cb3-SNAPSHOT which is before the fix for GenomicsDBImport
+                 *   sample name ordering was applied
+                 *
+                 *   the combined vcfs were then extracted using SelectVariants, these are the badlySorted1000 files
+                 *   each vcf contains a single record with 1000 genotypes that have a single format field "SN: sample name"
+                 *   these were created with SN's that matched their assigned sample names, but the badly sorted versions are shuffled
+                */
+
+                {getTestFile("reverseOrdered.sample_map"), getTestFile("badlySorted1000-batch-size-50.vcf"), 50},
+                {getTestFile("reverseOrdered.sample_map"), getTestFile("badlySorted1000-batch-size13.vcf"), 13},
+        };
+    }
+
+    @Test(dataProvider = "getBadlySortedFiles")
+    public void testUnshufflingRestoresCorrectSamples(final File sampleMap, final File shuffled, final int batchSize) throws IOException {
+        final ArgumentsBuilder args = new ArgumentsBuilder();
+        final File fixed = createTempFile("fixed_", ".vcf");
+        args.addVCF(shuffled)
+                .addFileArgument(GenomicsDBImport.SAMPLE_NAME_MAP_LONG_NAME, sampleMap)
+                .addArgument(GenomicsDBImport.BATCHSIZE_ARG_NAME, String.valueOf(batchSize))
+                .addOutput(fixed);
+        runCommandLine(args);
+
+        try(final FeatureReader<VariantContext> reader =
+                AbstractFeatureReader.getFeatureReader(fixed.getAbsolutePath(), new VCFCodec(), true)){
+            final VariantContext vc = reader.iterator().next();
+            vc.getGenotypes().iterator().forEachRemaining( g ->
+                    Assert.assertEquals(g.getSampleName(), g.getAnyAttribute(SAMPLE_NAME_KEY))
+            );
+        }
+
+    }
```

Comment:
This test seems very good. I'd also add a test proving that you don't improperly permute the existing sequence dictionary in the header (as I think you might currently be doing), and also some tests that you don't clobber vcfs that don't actually need fixing (eg., 5 samples and batchSize 50).

Reply:
Oh, and just thought of another test case: add one with a bgzipped/tabix-indexed vcf as input.

Reply:
removed the header sort

Reply:
adding a bgzipped vcf